src/main/java/com/zendesk/maxwell/producer/MaxwellKinesisProducer.java

```diff
@@ -32,82 +31,168 @@
+				UserRecordResult result = ((UserRecordFailedException) t).getResult();
+				List<String> errorList = new ArrayList<String>();
+				for(Attempt attempt : result.getAttempts()){
+				   errorList.add(String.format(
+						   "Delay after prev attempt: %d ms, "
+								   + "Duration: %d ms, Code: %s, "
+								   + "Message: %s",
+						   attempt.getDelay(), attempt.getDuration(),
+						   attempt.getErrorCode(),
+						   attempt.getErrorMessage()));
+				}
+				String errorListStr = StringUtils.join(errorList, "n");
+				logger.error(String.format(
+						"Record failed to put, attempts: %d. Errors: %s", result.getAttempts().size(), errorListStr));
+				logger.error(String.format("Number of failed records up to now: %d", errRecords));
+				logger.error(String.format("Errored message: %s", json));
+			}
+
+			cc.markCompleted();
+		} else {
+			attempts++;
 
-		logger.error("Exception during put", t);
+			logger.warn(String.format("Record failed to put. Retrying. Attempt: %d ", attempts));
 
-		cc.markCompleted();
+			try {
+				ByteBuffer encodedValue = ByteBuffer.wrap(json.getBytes("UTF-8"));
+				ListenableFuture<UserRecordResult> future = kinesisProducer.addUserRecord(kinesisStream, key, encodedValue);
+
+				FutureCallback<UserRecordResult> callback = new KinesisCallback(cc, position, key, json, successRecords, errorRecords, kinesisProducer, kinesisStream, attempts, maxAttempts);
+
+				Futures.addCallback(future, callback);
+			} catch (UnsupportedEncodingException e) {
+				logger.error("Error encoding message. Message: " + json + ". Error: " + e.getMessage());
+			}
+		}
 	};
 
 	@Override
 	public void onSuccess(UserRecordResult result) {
+		int succRecords = successRecords.incrementAndGet();
 		if(logger.isDebugEnabled()) {
-			logger.debug("->  key:" + key + ", shard id:" + result.getShardId() + ", sequence number:" + result.getSequenceNumber());
-			logger.debug("   " + json);
-			logger.debug("   " + position);
-			logger.debug("");
+			long totalTime = 0L;
+			for (Attempt attempt : result.getAttempts()) {
+				totalTime += attempt.getDelay() + attempt.getDuration();
+			}
+			logger.debug(String.format(
+					"Succesfully put record, sequenceNumber=%s, "
+							+ "shardId=%s, took %d attempts, "
+							+ "totalling %s ms",
+					result.getSequenceNumber(),
+					result.getShardId(), result.getAttempts().size(),
+					totalTime));
+			logger.debug(String.format("Number of success records up to now: %d", succRecords));
 		}
-
 		cc.markCompleted();
 	};
 }
 
-public class MaxwellKinesisProducer extends AbstractAsyncProducer {
+public class MaxwellKinesisProducer extends AbstractProducer {
+	private final ArrayBlockingQueue<RowMap> queue;
+	private final MaxwellKinesisProducerWorker worker;
+
+	public MaxwellKinesisProducer(MaxwellContext context, String kinesisStream) {
+		super(context);
+		this.queue = new ArrayBlockingQueue<>(100);
```

Comment:
@alejobs this is a good point since this will mean that the number of retries for a record wont be `maxAttempts` but instead `kplRetries * maxAttempts`

I don't have much opinion on this topic, but since the KPL does include retry logic, I personally wouldn't include it unless I encountered some kind issue and figured that the retry knobs that KPL exposes aren't enough to fix it.